### PR TITLE
Possible solution to cross version bug in blobs

### PR DESCRIPTION
**Pull request by dannimad**
For testing and showing purposes only. Tons of GC tests are failing after this change. 

There is a bug in current blob manager in which we resolve the blob handle using the localId. This is a bug because once we process the blob, we know the storageId (response.id - blobId (we need to unify concepts here)) and that should be used as the blob path. The reason for this is because 1.3.7 runtime version, does not know the concept of localId, so if we create a new container, create a blob in it and after that load an old container, the latter will be unable to process recognize the blob because it was processed using the storageId. This change is breaking tons of GC tests and need help debugging those.

### CODE REVIEW

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 1]
```diff
@@ -376,25 +376,33 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
 		);
 	}
 
-	private getBlobHandle(id: string): BlobHandle {
+	private getBlobHandle(localId: string, storageId: string | undefined): BlobHandle {
```
Comment (scottn12) [inline packages/runtime/container-runtime/src/blobManager.ts:379]: It's a little easier to make `storageId` optional - this way we never need to pass in `undefined` when we call `getBlobHandle()`
```suggestion
	private getBlobHandle(localId: string, storageId?: string): BlobHandle {
```

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 2]
```diff
@@ -376,25 +376,33 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
...
 			  }
 			: undefined;
-		return new BlobHandle(
-			`${BlobManager.basePath}/${id}`,
-			this.routeContext,
-			async () => this.getBlob(id),
-			callback,
-		);
+
+		return storageId
```
Comment (scottn12) [inline packages/runtime/container-runtime/src/blobManager.ts:393]: Better to explicitly check if it's undefined here
```suggestion
		return storageId !== undefined
```

**Review by scottn12 — COMMENTED** [review]
couple nits while we look into the GC issues

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 3]
```diff
@@ -404,7 +412,7 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
 		// The 'IDocumentStorageService.createBlob()' call below will respond with a localId.
 		const response = await this.getStorage().createBlob(blob);
 		this.setRedirection(response.id, undefined);
-		return this.getBlobHandle(response.id);
+		return this.getBlobHandle(response.id, undefined);
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/blobManager.ts:415]: why would be pass the remote id as the local id?

Comment (anthony-murphy) [reply packages/runtime/container-runtime/src/blobManager.ts:415]: this looks like a bug to me

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 4]
```diff
@@ -376,25 +376,33 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
...
-		return new BlobHandle(
-			`${BlobManager.basePath}/${id}`,
-			this.routeContext,
-			async () => this.getBlob(id),
-			callback,
-		);
+
+		return storageId
+			? new BlobHandle(
+					`${BlobManager.basePath}/${storageId}`,
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/blobManager.ts:395]: this seems wrong. we never validate the storage id is valid like we do the local id

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 3 above)
Comment (dannimad) [reply packages/runtime/container-runtime/src/blobManager.ts:415]: This is for detached scenario. Once attached we will have a "real" storageId. Basically creating a blob while detached creates a local id and behaves the same way as when stashing. You have a local only id which will later have a storage id

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 4 above)
Comment (dannimad) [reply packages/runtime/container-runtime/src/blobManager.ts:395]: Not sure what you mean by we never validate storage id

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 3 above)
Comment (dannimad) [reply packages/runtime/container-runtime/src/blobManager.ts:415]: And at the end it really won't have any difference in behaviour inside getBlobHandle whether we send it as the remote one. The change in getBlobHandle is basically just prioritize remote id over local id when we have both

**Comment by dannimad** [comment]
After talking with @agarwal-navin, patching 1.3.7 release makes more sense than this solution. He will describe the possible solution there and only if that fix doesn't work we will keep exploring solutions from 2.0.0 release.

**Pull request closed without merging**